### ISSUE TITLE
MdTextViewLineCollection.GetTextViewLineContaining* unoptimized

### ISSUE THREAD

**Issue by Therzok** [mono/monodevelop#4371]
`GetTextViewLineContainingBufferPosition` and `GetTextViewLineContainingYCoordinate` do linear scans for finding the line.

Lines are sorted, so:
1. verify out of bounds
2. then do a binary search

> VS bug [#591927](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/591927)

**Comment by Therzok** [comment]
See also vsts for additional info

**Comment by mkrueger** [comment]
Esp. we're running here into a .NET problem measuring java with 100  gives other results:
100 items:

linear search: 2542
binary search: 3538

200:

linear search: 3835
binary search: 3995

300:

linear search: 5148
binary search: 4510

```
import java.util.Random;

class Item
{
  public int Id;
  public Item(int id)
  {
    Id = id;
  }
}

class SearchTest
{
  static final int itemCount = 100;
  static final int count = 100000000;



  public static void main(String[] args)
  {
    Item[] items = new Item[itemCount];
    for (int i = 0; i < itemCount; i++)
      items[i] = new Item(i);
    long startTime = System.currentTimeMillis();
    Random rng = new Random(4096);
    for (int i = 0; i < count; i++)
    {
      int idx = rng.nextInt(itemCount - 1);
      for (int j = 0; j < itemCount; j++)
      {
        if (items[j].Id == idx)
          break;
      }
    }
    System.out.println("linear search: " + (System.currentTimeMillis() - startTime));
    rng = new Random(4096);
    startTime = System.currentTimeMillis();
    for (int i = 0; i < count; i++)
    {
        int idx = rng.nextInt(itemCount - 1);
        int lower = 0;
        int upper = itemCount;
        while (lower <= upper)
        {
            int mid = (lower + upper) / 2;
            int id = items[mid].Id;
            if (id < idx)
            {
                lower = mid + 1;
            }
            else if (id > idx)
            {
                upper = mid - 1;
            }
            else
            {
                break;
            }
        }
    }
    System.out.println("binary search: " + (System.currentTimeMillis() - startTime));
  }
}
```

**Comment by ToddGrun** [comment]
I think there's a flaw in your measurements. I actually tried out the VS implementation for GetTextViewLineContainingBufferPosition and I brought up completion a couple times to exercise this code. I had about 50 lines visible on screen with the completion coming up in about the middle of the screen. The binary search code was about 8 times faster.

Here's the code I tested:

		Stopwatch _swOld = new Stopwatch();
		Stopwatch _swNew = new Stopwatch();

		public ITextViewLine GetTextViewLineContainingBufferPosition (SnapshotPoint bufferPosition)
		{
			_swOld.Start ();
			ITextViewLine oldLine = this.FirstOrDefault (l => l.ContainsBufferPosition (bufferPosition));
			_swOld.Stop ();
			_swNew.Start ();
			int newLineIndex = this.FindTextViewLineIndexContainingBufferPosition (bufferPosition);
			ITextViewLine newLine = newLineIndex >= 0 ? this[newLineIndex] : null;
			_swNew.Stop ();

			return newLine;
		}

		private int FindTextViewLineIndexContainingBufferPosition(SnapshotPoint position)
		{
			if (position.Position < this[0].Start.Position)
				return -1;

			{
				//If the position starts on or after the start of the last line, then use the
				//line's ContainsBufferPosition logic to handle the special case at the end of
				//the buffer.
				ITextViewLine lastLine = this[this.Count - 1];

				if (position.Position >= lastLine.Start.Position)
					return lastLine.ContainsBufferPosition(position) ? (this.Count - 1) : -1;
			}

			int low = 0;
			int high = this.Count;
			while (low < high)
			{
				int middle = (low + high) / 2;
				ITextViewLine middleLine = this[middle];

				if (position.Position < middleLine.Start.Position)
					high = middle;
				else
					low = middle + 1;
			}

			return low - 1;
		}

**Comment by mkrueger** [comment]
No the measurements were correct but I didn't test LINQ.

I added:

```
            rng = new Random(4096);
            sw.Restart();
            for (int i = 0; i < count; i++)
            {
                var idx = rng.Next(itemCount - 1);
                items.FirstOrDefault(l => l.Id == idx);
            }
            sw.Stop();
            Console.WriteLine("LINQ search: " + sw.ElapsedMilliseconds);
```

Results for 50 items:

linear search:   5327
binary search:  4149
LINQ search:  30825

I knew that linq has a price but that's ridiculous - so that means: never use LINQ.

**Comment by Therzok** [comment]
Well, the problem there is that it's linq with captures. It becomes faster if it's a lambda without captures.

**Comment by mkrueger** [comment]
Not really

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace TestConsole2
{
    struct Item
    {
        public int Id { get; set; }
        public Item(int id)
        {
            Id = id;
        }
    }
    class MainClass
    {
        static int idx;
        static bool GetItem(Item item)
        {
            return item.Id == idx;
        }
        public static void Main(string[] args)
        {
            const int itemCount = 50;
            const int count = 100000000;
            List<Item> items = new List<Item>();
            for (int i = 0; i < itemCount; i++)
                items.Add(new Item(i));
            Stopwatch sw = new Stopwatch();
            sw.Start();
            var rng = new Random(4096);
            for (int i = 0; i < count; i++)
            {
                var idx = rng.Next(itemCount - 1);
                for (int j = 0; j < itemCount; j++)
                {
                    if (items[j].Id == idx)
                    {
                        break;
                    }
                }
            }
            sw.Stop();
            Console.WriteLine("linear search: " + sw.ElapsedMilliseconds);
            rng = new Random(4096);
            sw.Restart();
            for (int i = 0; i < count; i++)
            {
                var idx = rng.Next(itemCount - 1);
                var lower = 0;
                var upper = itemCount;
                while (lower <= upper)
                {
                    var mid = (lower + upper) / 2;
                    var id = items[mid].Id;
                    if (id < idx)
                    {
                        lower = mid + 1;
                    }
                    else if (id > idx)
                    {
                        upper = mid - 1;
                    }
                    else
                    {
                        break;
                    }
                }
            }
            sw.Stop();
            Console.WriteLine("binary search: " + sw.ElapsedMilliseconds);

            rng = new Random(4096);
            sw.Restart();
            for (int i = 0; i < count; i++)
            {
                var idx = rng.Next(itemCount - 1);
                items.FirstOrDefault(l => l.Id == idx);
            }
            sw.Stop();
            Console.WriteLine("LINQ search: " + sw.ElapsedMilliseconds);


            rng = new Random(4096);
            sw.Restart();
            for (int i = 0; i < count; i++)
            {
                MainClass.idx = rng.Next(itemCount - 1);
                items.FirstOrDefault(GetItem);
            }
            sw.Stop();
            Console.WriteLine("LINQ without captures search: " + sw.ElapsedMilliseconds);
        }
    }
}
```

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.6.0.463 (master)

Pull Request #4508 merged by: Mike Krüger
Author: mono
Commit: 4d8b7964934e36ef9f3c936f2833cfe40f5eb28e (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.5.0.1127 (d15-7)

Pull Request #4509 merged by: Mike Krüger
Author: mono
Commit: 63c88c6abd32e2b14368457e8b9f990090f45e3d (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.6.0.802 (master)

Pull Request #4508 merged by: Mike Kr??ger
Author: mono
Commit: 4d8b7964934e36ef9f3c936f2833cfe40f5eb28e (xamarin/monodevelop)
